Support binary mode for ICC

E.g. https://godbolt.org/g/86nUhZ - fails with missing libraries:

```
/opt/compiler-explorer/gcc-6.3.0/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../x86_64-linux-gnu/bin/ld: cannot find -lm
Compiler exited with result code 1
```

--- a/lib/base-compiler.js
+++ b/lib/base-compiler.js
@@ -305,7 +305,7 @@ Compile.prototype.compile = function (source, options, backendOptions, filters) 
                         asmResult.astOutput = astResult;
                     }
 
-                    if (gccDumpResult) {
+                    if (self.compiler.supportsGccDump && gccDumpResult) {
                         asmResult.hasGccDumpOutput = true;
                         asmResult.gccDumpOutput = gccDumpResult;
                     }

--- a/lib/compilers/argument-parsers.js
+++ b/lib/compilers/argument-parsers.js
@@ -48,6 +48,7 @@ const getOptions = function (compiler, helpArg) {
 
 const gccparser = function (compiler) {
     return getOptions(compiler, "--target-help").then(function (compiler) {
+        compiler.compiler.supportsGccDump = true;
         if (compiler.compiler.supportedOptions['-masm']) {
             compiler.compiler.intelAsm = "-masm=intel";
             compiler.compiler.supportsIntel = true;

--- a/static/compiler.js
+++ b/static/compiler.js
@@ -779,7 +779,7 @@ define(function (require) {
 
     Compiler.prototype.onGccDumpViewClosed = function (id) {
         if (this.id === id) {
-            this.gccDumpButton.prop('disabled', false);
+            this.gccDumpButton.prop('disabled', !this.compiler.supportsGccDump);
             this.gccDumpViewOpen = false;
 
             delete this.gccDumpPassSelected;
@@ -844,6 +844,12 @@ define(function (require) {
             this.cfgButton.prop('disabled', !this.compilerSupportsCfg);
         } else {
             this.cfgButton.prop('disabled', true);
+        }
+
+        if (!this.gccDumpViewOpen) {
+            this.gccDumpButton.prop('disabled', !this.compiler.supportsGccDump);
+        } else {
+            this.gccDumpButton.prop('disabled', true);
         }
     };
 

--- a/static/gccdump-view.js
+++ b/static/gccdump-view.js
@@ -50,7 +50,7 @@ define(function (require) {
         // disable UI callbacks.
         this.uiIsReady = false;
         // disable filter buttons
-        this.domRoot.find('.dump-filters').find('.btn')
+        this.domRoot.find('.dump-filters .btn')
             .each(function () {
                 $(this).addClass('disabled');
             });
@@ -127,13 +127,27 @@ define(function (require) {
         this.eventHub.emit('gccDumpUIInit', this.state._compilerid);
     }
 
+    // Disable view's menu when invalid compiler has been
+    // selected after view is opened.
+    GccDump.prototype.onUiNotReady = function () {
+        this.filters.off('change');
+        this.selectize.off('change');
+
+        // disable drop down menu and buttons
+        this.selectize.disable();
+        this.domRoot.find('.dump-filters .btn')
+            .each(function () {
+                $(this).addClass('disabled');
+            });
+    };
+
     GccDump.prototype.onUiReady = function () {
         this.filters.on('change', _.bind(this.onFilterChange, this));
         this.selectize.on('change', _.bind(this.onPassSelect, this));
 
         // enable drop down menu and buttons
         this.selectize.enable();
-        this.domRoot.find('.dump-filters').find('.btn')
+        this.domRoot.find('.dump-filters .btn')
             .each(function () {
                 $(this).removeClass('disabled');
             });
@@ -196,24 +210,6 @@ define(function (require) {
 
     GccDump.prototype.onCompileResult = function (id, compiler, result) {
         if (this.state._compilerid !== id) return;
-        // ignore spurious result when :
-        // - UI is disabled
-        // - some state has been restored
-        // - result does not have an gccdump info
-        //
-        // This happens when restoring a state and some builds are
-        // executed by another part of the code before we had the
-        // chance to set all our parameters : results are coming
-        // back without any dump info.
-        if (!this.uiIsReady &&
-            this.state.selectedPass !== '' &&
-            (!result.hasGccDumpOutput ||
-                result.gccDumpOutput.selectedPass === '')) {
-            // wait for correct build job:
-            // We should get something back from what has been
-            // requested earlier at the end of GccDump().
-            return;
-        }
 
         if (result.hasGccDumpOutput && result.gccDumpOutput.syntaxHighlight) {
             monaco.editor.setModelLanguage(this.gccDumpEditor.getModel(), 'gccdump-rtl-gimple');
@@ -232,16 +228,19 @@ define(function (require) {
             }
             this.updatePass(this.filters, this.selectize, result.gccDumpOutput);
             this.showGccDumpResults(currOutput);
+
+            // enable UI on first successful compilation or after an invalid compiler selection (eg. clang)
+            if (!this.uiIsReady) {
+                this.uiIsReady = true;
+                this.onUiReady();
+            }
         } else {
             this.selectize.clear(true);
             this.state.selectedPass = '';
             this.updatePass(this.filters, this.selectize, false);
-            this.showGccDumpResults('<no output>');
-        }
-        // enable UI on first compilation
-        if (!this.uiIsReady) {
-            this.uiIsReady = true;
-            this.onUiReady();
+            this.showGccDumpResults('<No output, maybe you switched to a non-GCC compiler ?>');
+            this.uiIsReady = false;
+            this.onUiNotReady();
         }
 
         this.saveState();

--- a/views/templates.pug
+++ b/views/templates.pug
@@ -53,7 +53,7 @@
             span.glyphicon.glyphicon-scale
           button.btn.btn-default.btn-sm.view-ast(title="Show AST output (Clang only)")
             span.glyphicon.glyphicon-tree-deciduous
-          button.btn.btn-default.btn-sm.view-gccdump(title="Show GCC's Tree/RTL dump")
+          button.btn.btn-default.btn-sm.view-gccdump(title="Show Tree/RTL dump (GCC only)")
             span.glyphicon.glyphicon-tree-conifer
           button.btn.btn-default.btn-sm.view-cfg(title="Show Graph Output")
             span.glyphicon.glyphicon-lamp

